## Import libraries


In [1]:
import numpy as np
import pandas as pd
from random import choice, uniform, random, randint
from operator import itemgetter, attrgetter
import keras
from keras import backend as K
from keras.models import Sequential
from keras.layers import Conv2D,MaxPooling2D,Dropout,Activation,Flatten,Dense
from keras.optimizers import Adam
from keras.preprocessing.image import ImageDataGenerator
import os
from tensorflow.keras import backend
from tensorflow.keras.applications import VGG16, MobileNet
import numpy as np
import pandas as pd
import os
import matplotlib.pyplot as plt
from IPython.display import Image, display
import random
import math
import tensorflow as tf
from keras.preprocessing.image import load_img,img_to_array,ImageDataGenerator
from tensorflow.keras.models import Model, Sequential
from tensorflow.keras.utils import plot_model 
from tensorflow.keras.layers import Input, Dense, Dropout, Flatten, Activation,Concatenate
from tensorflow.keras.layers import Conv2D, MaxPooling2D, AveragePooling2D
from tensorflow.keras.layers import BatchNormalization
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.optimizers import Adam, SGD
from tensorflow.keras.callbacks import EarlyStopping, LearningRateScheduler, ReduceLROnPlateau
from tensorflow.keras import backend, models
from sklearn.metrics import confusion_matrix
from tensorflow.keras.applications import VGG16, MobileNet
from keras.applications.vgg16 import preprocess_input

In [2]:
population = []
noPop = 6
fmap = [16,32,64,128,256]
kernel = [2,3,5,7,9]
activation = ['relu','sigmoid','tanh']
dropout = [0.0,0.1,0.2,0.3,0.4,0.5]
learningRate = [0.0001,0.0005, 0.001, 0.005, 0.01,0.05,0.1,0.5] 
mutationRate = 0.3 
noGen = 10 

## Preparation and analysis of dataset

In [3]:
np.random.seed(12)
train = ImageDataGenerator(rescale=1./255).flow_from_directory('dataset/train',target_size=(224,224))

Found 35215 images belonging to 250 classes.


## Genetic Algorithm

In [4]:
def initializingPopulation(population,noPop):
    for i in range (0,noPop):
        loss = 0.0
        accuracy = 0.0
        inPop = [choice(fmap),choice(kernel),choice(activation),choice(dropout),choice(learningRate),loss,accuracy]
        population.append (inPop)

def rankSelection (population,noPop): 
    population = sorted(population,key=itemgetter(5))
    population = sorted(population,key=itemgetter(6),reverse=True)
    noHalfPopulation = noPop//2 
    halfPopulation = population[:noHalfPopulation]

    return halfPopulation

def tournamentSelection(selectedRank,tSize):
    best = None 
    for i in range(0,tSize):
        ind = choice(selectedRank)
        if best == None:
            best = ind
        elif best == ind:
            ind = choice(selectedRank)

        if best[5]>ind[5] and best[6]<ind[6]:
            best = ind

    return best

def crossover(p1,p2): 
    crossoverPoint = len(p1)//2 
    c1 = p1[:crossoverPoint] + p2[crossoverPoint:]
    c2 = p2[:crossoverPoint] + p1[crossoverPoint:]
    return c1, c2

def mutation (c1,c2):
    rate = random.random()
    print(c1)
    print(c2)
    
    print ('rate: ' +str(rate))
    if rate<=mutationRate:
        length = len(c1)-2
        print ('length: '+str(length))
        index = randint(0,length)
        print('index: ' +str(index))

        if index == 0:
            c1[index] = choice(fmap)
            c2[index] = choice(fmap)
        elif index == 1:
            c1[index] = choice(kernel)
            c2[index] = choice(kernel)
        elif index == 2:
            c1[index] = choice(activation)
            c2[index] = choice(activation)
        elif index == 3:
            c1[index] = choice(dropout)
            c2[index] = choice(dropout)
        elif index == 4:
            c1[index] = choice(learningRate)
            c2[index] = choice(learningRate)
    return c1,c2

In [5]:
def cnn(population):
    for p in population:
        #this is where the fitness value will evaluate
        f = p[0] # features map variable
        k = p[1] # kernel variable
        a = p[2] # activation function variable
        d = p[3] # dropout variable
        l = p[4] # learning rate variable
        print(f,k,a,d,l)
        
        backend.clear_session()

        base_mobilenet = MobileNet(weights = 'imagenet', include_top = False, 
                           input_shape = (224,224,3))
        base_mobilenet.trainable = False # Freeze the mobilenet weights.

        model = Sequential()
        model.add(base_mobilenet)
        model.add(Conv2D(f,(k,k),input_shape=(224,224,3),padding='same'))
        model.add(Activation(a))
        model.add(MaxPooling2D(pool_size=(2,2),padding='same'))
        model.add(Flatten())
        model.add(Dense(32*32))
        model.add(Dropout(d))
        model.add(Dense(250))
        model.add(Activation('softmax'))

        model.compile(Adam(l),loss='categorical_crossentropy',metrics=['accuracy'])

        history = model.fit(train,epochs=100,steps_per_epoch=10,verbose=2)
        acc = np.mean(history.history['accuracy'])
        loss = np.mean(history.history['loss'])
        print(acc)
        print(loss)

        p[5] = loss
        p[6] = acc
        print(p[5])
        print(p[6])


initializingPopulation(population,noPop)

print ('Generation 0')
for p in population:
    print(p)

for gen in range(0,noGen):

    cnn(population)

    selectedRank = rankSelection(population,noPop)
    print('Selected Ranked Chromosomes:')
    for sr in selectedRank:
        print(sr)

    if selectedRank[-1][5]<6 and selectedRank[-1][6]>0.6:
        # this is  stopping condition for genetic algorithm.
        # you can change or delete it if you want
        # but, remember, waiting for 10 generation is time consumption
        # and also computationally overhead
        break
    else:
        child = [] 
        tSize = 2 
        numberOfMating = len(selectedRank)//2
        # this is limiter  for number of child will generate
        # let say, you have 12 total population
        # let say, you have 6 ranking selection
        # the mating process will be 3 time
        # but, since the generated child is 2
        # thus mating process would generate 6 children
        # this is because number of mating * generated Children = half new population
        # 3 mating process * 2 newly generated child = 6 population
        for i in range (0,numberOfMating):
            #this is where mating process take place
            parent1 = tournamentSelection(selectedRank,tSize)
            parent2 = tournamentSelection(selectedRank,tSize)
            while parent2 is parent1:
                parent2 = tournamentSelection(selectedRank,tSize)
                #this is to make sure the candidate does not identical
            c1,c2 = crossover(parent1,parent2)
            print('Child 1: %s child 2: %s'%(c1,c2))
            c1,c2 = mutation(c1,c2)
            child.append(c1)
            child.append(c2)

        population.clear() 
        population = selectedRank + child 

        print ('New Child' + str(gen))
        for c in child:
            print(c)

        print('New Generation ' + str(gen+1))
        for p in population:
            print(p)

Generation 0
[64, 2, 'sigmoid', 0.5, 0.001, 0.0, 0.0]
[256, 2, 'tanh', 0.5, 0.001, 0.0, 0.0]
[256, 9, 'tanh', 0.0, 0.5, 0.0, 0.0]
[256, 5, 'tanh', 0.3, 0.1, 0.0, 0.0]
[32, 9, 'relu', 0.3, 0.01, 0.0, 0.0]
[128, 7, 'sigmoid', 0.2, 0.1, 0.0, 0.0]
64 2 sigmoid 0.5 0.001
Epoch 1/100
10/10 - 10s - loss: 6.3604 - accuracy: 0.0031
Epoch 2/100
10/10 - 8s - loss: 5.7967 - accuracy: 0.0000e+00
Epoch 3/100
10/10 - 7s - loss: 5.6206 - accuracy: 0.0063
Epoch 4/100
10/10 - 8s - loss: 5.6348 - accuracy: 0.0125
Epoch 5/100
10/10 - 7s - loss: 5.5890 - accuracy: 0.0063
Epoch 6/100
10/10 - 7s - loss: 5.5827 - accuracy: 0.0031
Epoch 7/100
10/10 - 7s - loss: 5.5150 - accuracy: 0.0094
Epoch 8/100
10/10 - 7s - loss: 5.5435 - accuracy: 0.0094
Epoch 9/100
10/10 - 7s - loss: 5.5472 - accuracy: 0.0156
Epoch 10/100
10/10 - 7s - loss: 5.4900 - accuracy: 0.0188
Epoch 11/100
10/10 - 7s - loss: 5.5431 - accuracy: 0.0156
Epoch 12/100
10/10 - 7s - loss: 5.4589 - accuracy: 0.0094
Epoch 13/100
10/10 - 7s - loss: 5.4651 - 

10/10 - 7s - loss: 5.3178 - accuracy: 0.3531
Epoch 37/100
10/10 - 8s - loss: 5.2959 - accuracy: 0.3438
Epoch 38/100
10/10 - 7s - loss: 5.0830 - accuracy: 0.3750
Epoch 39/100
10/10 - 7s - loss: 4.5381 - accuracy: 0.4031
Epoch 40/100
10/10 - 7s - loss: 4.3717 - accuracy: 0.4219
Epoch 41/100
10/10 - 7s - loss: 4.7286 - accuracy: 0.3719
Epoch 42/100
10/10 - 7s - loss: 4.5048 - accuracy: 0.4062
Epoch 43/100
10/10 - 7s - loss: 4.6196 - accuracy: 0.4156
Epoch 44/100
10/10 - 7s - loss: 4.8405 - accuracy: 0.4156
Epoch 45/100
10/10 - 8s - loss: 4.3166 - accuracy: 0.4625
Epoch 46/100
10/10 - 7s - loss: 4.2705 - accuracy: 0.4875
Epoch 47/100
10/10 - 7s - loss: 4.3365 - accuracy: 0.4375
Epoch 48/100
10/10 - 9s - loss: 4.4058 - accuracy: 0.4375
Epoch 49/100
10/10 - 9s - loss: 4.1087 - accuracy: 0.4750
Epoch 50/100
10/10 - 9s - loss: 4.7971 - accuracy: 0.4469
Epoch 51/100
10/10 - 9s - loss: 4.8531 - accuracy: 0.4281
Epoch 52/100
10/10 - 8s - loss: 4.5996 - accuracy: 0.4344
Epoch 53/100
10/10 - 8s - l

Epoch 67/100
10/10 - 24s - loss: 10025777.0000 - accuracy: 0.0031
Epoch 68/100
10/10 - 22s - loss: 7969865.5000 - accuracy: 0.0094
Epoch 69/100
10/10 - 22s - loss: 8080796.0000 - accuracy: 0.0063
Epoch 70/100
10/10 - 22s - loss: 8234127.0000 - accuracy: 0.0000e+00
Epoch 71/100
10/10 - 22s - loss: 8739195.0000 - accuracy: 0.0063
Epoch 72/100
10/10 - 22s - loss: 9626519.0000 - accuracy: 0.0000e+00
Epoch 73/100
10/10 - 25s - loss: 8618994.0000 - accuracy: 0.0000e+00
Epoch 74/100
10/10 - 23s - loss: 9706897.0000 - accuracy: 0.0031
Epoch 75/100
10/10 - 22s - loss: 8829172.0000 - accuracy: 0.0063
Epoch 76/100
10/10 - 22s - loss: 10446041.0000 - accuracy: 0.0094
Epoch 77/100
10/10 - 23s - loss: 9065094.0000 - accuracy: 0.0000e+00
Epoch 78/100
10/10 - 23s - loss: 8875594.0000 - accuracy: 0.0031
Epoch 79/100
10/10 - 22s - loss: 7656745.5000 - accuracy: 0.0031
Epoch 80/100
10/10 - 24s - loss: 7449342.5000 - accuracy: 0.0063
Epoch 81/100
10/10 - 24s - loss: 8144105.0000 - accuracy: 0.0000e+00
Epo

Epoch 93/100
10/10 - 10s - loss: 100315.7969 - accuracy: 0.0063
Epoch 94/100
10/10 - 10s - loss: 101340.8594 - accuracy: 0.0125
Epoch 95/100
10/10 - 11s - loss: 90734.9922 - accuracy: 0.0031
Epoch 96/100
10/10 - 12s - loss: 86125.7891 - accuracy: 0.0063
Epoch 97/100
10/10 - 11s - loss: 81472.7891 - accuracy: 0.0156
Epoch 98/100
10/10 - 10s - loss: 90496.3984 - accuracy: 0.0063
Epoch 99/100
10/10 - 10s - loss: 88869.7969 - accuracy: 0.0094
Epoch 100/100
10/10 - 10s - loss: 95442.4844 - accuracy: 0.0031
0.005000000102445484
88517.8793359375
88517.8793359375
0.005000000102445484
32 9 relu 0.3 0.01
Epoch 1/100
10/10 - 10s - loss: 34.5584 - accuracy: 0.0125
Epoch 2/100
10/10 - 9s - loss: 5.6380 - accuracy: 0.0000e+00
Epoch 3/100
10/10 - 9s - loss: 5.5630 - accuracy: 0.0031
Epoch 4/100
10/10 - 9s - loss: 5.5872 - accuracy: 0.0063
Epoch 5/100
10/10 - 9s - loss: 5.7195 - accuracy: 0.0000e+00
Epoch 6/100
10/10 - 9s - loss: 5.5556 - accuracy: 0.0219
Epoch 7/100
10/10 - 9s - loss: 5.6056 - accura

Epoch 27/100
10/10 - 12s - loss: 1866.4189 - accuracy: 0.0031
Epoch 28/100
10/10 - 11s - loss: 1989.2494 - accuracy: 0.0063
Epoch 29/100
10/10 - 10s - loss: 1925.3802 - accuracy: 0.0031
Epoch 30/100
10/10 - 10s - loss: 2109.8594 - accuracy: 0.0094
Epoch 31/100
10/10 - 10s - loss: 1774.9395 - accuracy: 0.0063
Epoch 32/100
10/10 - 10s - loss: 2112.0222 - accuracy: 0.0031
Epoch 33/100
10/10 - 10s - loss: 2292.2903 - accuracy: 0.0063
Epoch 34/100
10/10 - 10s - loss: 2885.8552 - accuracy: 0.0125
Epoch 35/100
10/10 - 10s - loss: 2698.8875 - accuracy: 0.0031
Epoch 36/100
10/10 - 10s - loss: 2492.3188 - accuracy: 0.0094
Epoch 37/100
10/10 - 10s - loss: 2419.5913 - accuracy: 0.0063
Epoch 38/100
10/10 - 10s - loss: 2277.0391 - accuracy: 0.0031
Epoch 39/100
10/10 - 10s - loss: 2298.4341 - accuracy: 0.0000e+00
Epoch 40/100
10/10 - 10s - loss: 3047.9487 - accuracy: 0.0063
Epoch 41/100
10/10 - 10s - loss: 2641.6572 - accuracy: 0.0094
Epoch 42/100
10/10 - 10s - loss: 3015.5801 - accuracy: 0.0000e+00


Epoch 42/100
10/10 - 7s - loss: 4.2331 - accuracy: 0.4563
Epoch 43/100
10/10 - 7s - loss: 3.7966 - accuracy: 0.5031
Epoch 44/100
10/10 - 7s - loss: 4.2185 - accuracy: 0.4500
Epoch 45/100
10/10 - 7s - loss: 4.4179 - accuracy: 0.4219
Epoch 46/100
10/10 - 7s - loss: 4.1851 - accuracy: 0.4281
Epoch 47/100
10/10 - 7s - loss: 4.6520 - accuracy: 0.4469
Epoch 48/100
10/10 - 7s - loss: 4.3937 - accuracy: 0.4344
Epoch 49/100
10/10 - 7s - loss: 3.9836 - accuracy: 0.4594
Epoch 50/100
10/10 - 7s - loss: 4.6413 - accuracy: 0.4719
Epoch 51/100
10/10 - 7s - loss: 4.9260 - accuracy: 0.4250
Epoch 52/100
10/10 - 7s - loss: 4.4338 - accuracy: 0.4781
Epoch 53/100
10/10 - 7s - loss: 4.6396 - accuracy: 0.4750
Epoch 54/100
10/10 - 7s - loss: 3.9764 - accuracy: 0.5188
Epoch 55/100
10/10 - 6s - loss: 3.9490 - accuracy: 0.4752
Epoch 56/100
10/10 - 7s - loss: 3.9097 - accuracy: 0.5125
Epoch 57/100
10/10 - 7s - loss: 3.5315 - accuracy: 0.5531
Epoch 58/100
10/10 - 7s - loss: 3.7426 - accuracy: 0.5437
Epoch 59/100
1

Epoch 82/100
10/10 - 6s - loss: 2.0682 - accuracy: 0.4906
Epoch 83/100
10/10 - 6s - loss: 2.1263 - accuracy: 0.4781
Epoch 84/100
10/10 - 6s - loss: 1.9449 - accuracy: 0.4969
Epoch 85/100
10/10 - 6s - loss: 1.8608 - accuracy: 0.4844
Epoch 86/100
10/10 - 6s - loss: 2.0009 - accuracy: 0.4781
Epoch 87/100
10/10 - 6s - loss: 1.9876 - accuracy: 0.4781
Epoch 88/100
10/10 - 6s - loss: 2.0575 - accuracy: 0.4938
Epoch 89/100
10/10 - 6s - loss: 2.1820 - accuracy: 0.4375
Epoch 90/100
10/10 - 6s - loss: 1.8846 - accuracy: 0.4969
Epoch 91/100
10/10 - 6s - loss: 1.7470 - accuracy: 0.5562
Epoch 92/100
10/10 - 6s - loss: 1.7581 - accuracy: 0.5406
Epoch 93/100
10/10 - 6s - loss: 1.7381 - accuracy: 0.5469
Epoch 94/100
10/10 - 6s - loss: 1.7401 - accuracy: 0.5469
Epoch 95/100
10/10 - 6s - loss: 1.7527 - accuracy: 0.4938
Epoch 96/100
10/10 - 6s - loss: 1.6126 - accuracy: 0.5688
Epoch 97/100
10/10 - 6s - loss: 1.6388 - accuracy: 0.5656
Epoch 98/100
10/10 - 6s - loss: 1.7075 - accuracy: 0.5437
Epoch 99/100
1

10/10 - 6s - loss: 6.7336 - accuracy: 0.0000e+00
Epoch 13/100
10/10 - 6s - loss: 7.3821 - accuracy: 0.0031
Epoch 14/100
10/10 - 6s - loss: 7.3518 - accuracy: 0.0000e+00
Epoch 15/100
10/10 - 6s - loss: 6.7416 - accuracy: 0.0156
Epoch 16/100
10/10 - 6s - loss: 7.1328 - accuracy: 0.0000e+00
Epoch 17/100
10/10 - 6s - loss: 6.9665 - accuracy: 0.0000e+00
Epoch 18/100
10/10 - 6s - loss: 7.2056 - accuracy: 0.0031
Epoch 19/100
10/10 - 6s - loss: 6.7365 - accuracy: 0.0000e+00
Epoch 20/100
10/10 - 6s - loss: 7.0172 - accuracy: 0.0094
Epoch 21/100
10/10 - 6s - loss: 6.7838 - accuracy: 0.0031
Epoch 22/100
10/10 - 6s - loss: 7.1806 - accuracy: 0.0000e+00
Epoch 23/100
10/10 - 6s - loss: 6.7528 - accuracy: 0.0000e+00
Epoch 24/100
10/10 - 6s - loss: 7.0309 - accuracy: 0.0000e+00
Epoch 25/100
10/10 - 6s - loss: 6.7695 - accuracy: 0.0188
Epoch 26/100
10/10 - 7s - loss: 7.0760 - accuracy: 0.0031
Epoch 27/100
10/10 - 6s - loss: 7.0625 - accuracy: 0.0094
Epoch 28/100
10/10 - 6s - loss: 7.2805 - accuracy: 0.

Epoch 50/100
10/10 - 7s - loss: 4.2892 - accuracy: 0.4688
Epoch 51/100
10/10 - 7s - loss: 4.4228 - accuracy: 0.4406
Epoch 52/100
10/10 - 7s - loss: 3.9656 - accuracy: 0.4844
Epoch 53/100
10/10 - 7s - loss: 4.1233 - accuracy: 0.4812
Epoch 54/100
10/10 - 7s - loss: 4.1368 - accuracy: 0.4844
Epoch 55/100
10/10 - 7s - loss: 4.5550 - accuracy: 0.4875
Epoch 56/100
10/10 - 7s - loss: 4.7033 - accuracy: 0.4594
Epoch 57/100
10/10 - 7s - loss: 5.1749 - accuracy: 0.4469
Epoch 58/100
10/10 - 7s - loss: 4.5543 - accuracy: 0.4563
Epoch 59/100
10/10 - 7s - loss: 3.9558 - accuracy: 0.5094
Epoch 60/100
10/10 - 7s - loss: 4.0871 - accuracy: 0.5156
Epoch 61/100
10/10 - 7s - loss: 4.3787 - accuracy: 0.4719
Epoch 62/100
10/10 - 7s - loss: 4.4947 - accuracy: 0.4750
Epoch 63/100
10/10 - 7s - loss: 4.0455 - accuracy: 0.5063
Epoch 64/100
10/10 - 7s - loss: 4.6037 - accuracy: 0.5156
Epoch 65/100
10/10 - 7s - loss: 4.2346 - accuracy: 0.5063
Epoch 66/100
10/10 - 7s - loss: 4.2647 - accuracy: 0.5344
Epoch 67/100
1

Epoch 71/100
10/10 - 8s - loss: 4.0178 - accuracy: 0.5125
Epoch 72/100
10/10 - 8s - loss: 3.5356 - accuracy: 0.5344
Epoch 73/100
10/10 - 8s - loss: 4.0453 - accuracy: 0.5344
Epoch 74/100
10/10 - 8s - loss: 3.8283 - accuracy: 0.5531
Epoch 75/100
10/10 - 8s - loss: 3.3796 - accuracy: 0.5625
Epoch 76/100
10/10 - 8s - loss: 3.7205 - accuracy: 0.5719
Epoch 77/100
10/10 - 8s - loss: 3.6026 - accuracy: 0.5781
Epoch 78/100
10/10 - 8s - loss: 3.8022 - accuracy: 0.5719
Epoch 79/100
10/10 - 8s - loss: 3.9432 - accuracy: 0.5094
Epoch 80/100
10/10 - 8s - loss: 4.4516 - accuracy: 0.5344
Epoch 81/100
10/10 - 8s - loss: 4.1913 - accuracy: 0.5531
Epoch 82/100
10/10 - 8s - loss: 3.7275 - accuracy: 0.5688
Epoch 83/100
10/10 - 8s - loss: 3.6189 - accuracy: 0.6031
Epoch 84/100
10/10 - 8s - loss: 4.3055 - accuracy: 0.5688
Epoch 85/100
10/10 - 8s - loss: 3.7478 - accuracy: 0.5719
Epoch 86/100
10/10 - 8s - loss: 4.4745 - accuracy: 0.5688
Epoch 87/100
10/10 - 8s - loss: 3.6454 - accuracy: 0.5719
Epoch 88/100
1

10/10 - 8s - loss: 5.5441 - accuracy: 0.0156
Epoch 10/100
10/10 - 8s - loss: 5.5208 - accuracy: 0.0063
Epoch 11/100
10/10 - 7s - loss: 5.5001 - accuracy: 0.0094
Epoch 12/100
10/10 - 8s - loss: 5.4472 - accuracy: 0.0125
Epoch 13/100
10/10 - 8s - loss: 5.4709 - accuracy: 0.0031
Epoch 14/100
10/10 - 8s - loss: 5.3931 - accuracy: 0.0094
Epoch 15/100
10/10 - 8s - loss: 5.3939 - accuracy: 0.0156
Epoch 16/100
10/10 - 8s - loss: 5.4052 - accuracy: 0.0219
Epoch 17/100
10/10 - 8s - loss: 5.3221 - accuracy: 0.0094
Epoch 18/100
10/10 - 8s - loss: 5.3035 - accuracy: 0.0219
Epoch 19/100
10/10 - 8s - loss: 5.1331 - accuracy: 0.0500
Epoch 20/100
10/10 - 8s - loss: 5.2646 - accuracy: 0.0188
Epoch 21/100
10/10 - 8s - loss: 5.1233 - accuracy: 0.0281
Epoch 22/100
10/10 - 8s - loss: 5.0843 - accuracy: 0.0250
Epoch 23/100
10/10 - 8s - loss: 5.0657 - accuracy: 0.0406
Epoch 24/100
10/10 - 8s - loss: 5.0217 - accuracy: 0.0312
Epoch 25/100
10/10 - 8s - loss: 5.0381 - accuracy: 0.0281
Epoch 26/100
10/10 - 8s - l

Epoch 49/100
10/10 - 8s - loss: 3.3779 - accuracy: 0.2156
Epoch 50/100
10/10 - 8s - loss: 3.5104 - accuracy: 0.2031
Epoch 51/100
10/10 - 8s - loss: 3.4308 - accuracy: 0.2125
Epoch 52/100
10/10 - 8s - loss: 3.3403 - accuracy: 0.2281
Epoch 53/100
10/10 - 8s - loss: 3.3256 - accuracy: 0.2469
Epoch 54/100
10/10 - 8s - loss: 3.0965 - accuracy: 0.2656
Epoch 55/100
10/10 - 8s - loss: 3.0624 - accuracy: 0.2875
Epoch 56/100
10/10 - 8s - loss: 3.0633 - accuracy: 0.2594
Epoch 57/100
10/10 - 8s - loss: 3.1045 - accuracy: 0.2562
Epoch 58/100
10/10 - 8s - loss: 2.9356 - accuracy: 0.3219
Epoch 59/100
10/10 - 8s - loss: 2.9472 - accuracy: 0.2906
Epoch 60/100
10/10 - 8s - loss: 3.0374 - accuracy: 0.2531
Epoch 61/100
10/10 - 8s - loss: 2.8886 - accuracy: 0.3219
Epoch 62/100
10/10 - 8s - loss: 2.9015 - accuracy: 0.2594
Epoch 63/100
10/10 - 8s - loss: 2.8235 - accuracy: 0.3125
Epoch 64/100
10/10 - 8s - loss: 2.7870 - accuracy: 0.3313
Epoch 65/100
10/10 - 8s - loss: 2.8186 - accuracy: 0.3187
Epoch 66/100
1

Epoch 89/100
10/10 - 8s - loss: 1.5904 - accuracy: 0.6156
Epoch 90/100
10/10 - 8s - loss: 1.5592 - accuracy: 0.5969
Epoch 91/100
10/10 - 8s - loss: 1.7266 - accuracy: 0.5281
Epoch 92/100
10/10 - 8s - loss: 1.4814 - accuracy: 0.6062
Epoch 93/100
10/10 - 8s - loss: 1.6283 - accuracy: 0.5906
Epoch 94/100
10/10 - 8s - loss: 1.8544 - accuracy: 0.5469
Epoch 95/100
10/10 - 8s - loss: 1.5257 - accuracy: 0.6281
Epoch 96/100
10/10 - 8s - loss: 1.6469 - accuracy: 0.6062
Epoch 97/100
10/10 - 8s - loss: 1.8057 - accuracy: 0.5719
Epoch 98/100
10/10 - 8s - loss: 1.4070 - accuracy: 0.6469
Epoch 99/100
10/10 - 8s - loss: 1.6851 - accuracy: 0.5906
Epoch 100/100
10/10 - 8s - loss: 1.6319 - accuracy: 0.5969
0.40775979732628914
2.6197396981716157
2.6197396981716157
0.40775979732628914
Selected Ranked Chromosomes:
[16, 2, 'tanh', 0.5, 0.001, 2.6197396981716157, 0.40775979732628914]
[256, 2, 'tanh', 0.5, 0.001, 5.215166244506836, 0.39575824967585504]
[256, 2, 'tanh', 0.5, 0.001, 5.243131053447724, 0.39043750

Epoch 9/100
10/10 - 8s - loss: 9.0586 - accuracy: 0.0500
Epoch 10/100
10/10 - 8s - loss: 8.4287 - accuracy: 0.1031
Epoch 11/100
10/10 - 8s - loss: 8.8005 - accuracy: 0.0875
Epoch 12/100
10/10 - 8s - loss: 8.9133 - accuracy: 0.0625
Epoch 13/100
10/10 - 9s - loss: 8.3288 - accuracy: 0.1219
Epoch 14/100
10/10 - 8s - loss: 7.2787 - accuracy: 0.1250
Epoch 15/100
10/10 - 8s - loss: 7.4992 - accuracy: 0.1063
Epoch 16/100
10/10 - 8s - loss: 7.5011 - accuracy: 0.1312
Epoch 17/100
10/10 - 8s - loss: 7.2327 - accuracy: 0.1656
Epoch 18/100
10/10 - 8s - loss: 7.1204 - accuracy: 0.1562
Epoch 19/100
10/10 - 8s - loss: 6.2707 - accuracy: 0.2156
Epoch 20/100
10/10 - 8s - loss: 6.8437 - accuracy: 0.1688
Epoch 21/100
10/10 - 8s - loss: 5.9000 - accuracy: 0.2469
Epoch 22/100
10/10 - 8s - loss: 6.0247 - accuracy: 0.2625
Epoch 23/100
10/10 - 8s - loss: 6.0157 - accuracy: 0.2125
Epoch 24/100
10/10 - 8s - loss: 6.0859 - accuracy: 0.2688
Epoch 25/100
10/10 - 8s - loss: 6.0363 - accuracy: 0.2781
Epoch 26/100
10

Epoch 49/100
10/10 - 8s - loss: 4.0369 - accuracy: 0.4719
Epoch 50/100
10/10 - 8s - loss: 5.1013 - accuracy: 0.4031
Epoch 51/100
10/10 - 8s - loss: 4.6751 - accuracy: 0.4156
Epoch 52/100
10/10 - 8s - loss: 4.6234 - accuracy: 0.4531
Epoch 53/100
10/10 - 8s - loss: 4.3524 - accuracy: 0.5000
Epoch 54/100
10/10 - 8s - loss: 3.8555 - accuracy: 0.5281
Epoch 55/100
10/10 - 8s - loss: 4.2481 - accuracy: 0.4812
Epoch 56/100
10/10 - 8s - loss: 3.9974 - accuracy: 0.5281
Epoch 57/100
10/10 - 8s - loss: 4.4948 - accuracy: 0.4875
Epoch 58/100
10/10 - 8s - loss: 4.0262 - accuracy: 0.5063
Epoch 59/100
10/10 - 8s - loss: 4.2569 - accuracy: 0.4969
Epoch 60/100
10/10 - 8s - loss: 3.8247 - accuracy: 0.5219
Epoch 61/100
10/10 - 8s - loss: 3.9473 - accuracy: 0.5156
Epoch 62/100
10/10 - 8s - loss: 4.4522 - accuracy: 0.4719
Epoch 63/100
10/10 - 8s - loss: 4.7687 - accuracy: 0.4844
Epoch 64/100
10/10 - 8s - loss: 3.6231 - accuracy: 0.5250
Epoch 65/100
10/10 - 8s - loss: 4.2057 - accuracy: 0.4906
Epoch 66/100
1

Epoch 89/100
10/10 - 8s - loss: 1.8624 - accuracy: 0.5406
Epoch 90/100
10/10 - 8s - loss: 1.6985 - accuracy: 0.5500
Epoch 91/100
10/10 - 8s - loss: 1.7217 - accuracy: 0.5594
Epoch 92/100
10/10 - 8s - loss: 1.5958 - accuracy: 0.5906
Epoch 93/100
10/10 - 8s - loss: 1.6242 - accuracy: 0.5969
Epoch 94/100
10/10 - 8s - loss: 1.7573 - accuracy: 0.5688
Epoch 95/100
10/10 - 7s - loss: 1.5854 - accuracy: 0.6094
Epoch 96/100
10/10 - 8s - loss: 1.9609 - accuracy: 0.4938
Epoch 97/100
10/10 - 8s - loss: 1.5227 - accuracy: 0.5844
Epoch 98/100
10/10 - 8s - loss: 1.5491 - accuracy: 0.5844
Epoch 99/100
10/10 - 8s - loss: 1.6376 - accuracy: 0.6031
Epoch 100/100
10/10 - 8s - loss: 1.7479 - accuracy: 0.5344
0.3997030736738816
2.662250417470932
2.662250417470932
0.3997030736738816
256 2 tanh 0.5 0.001
Epoch 1/100
10/10 - 9s - loss: 9.5470 - accuracy: 0.0031
Epoch 2/100
10/10 - 8s - loss: 10.5599 - accuracy: 0.0250
Epoch 3/100
10/10 - 8s - loss: 10.2573 - accuracy: 0.0094
Epoch 4/100
10/10 - 8s - loss: 9.90

Epoch 9/100
10/10 - 8s - loss: 8.6616 - accuracy: 0.0562
Epoch 10/100
10/10 - 8s - loss: 8.7612 - accuracy: 0.0531
Epoch 11/100
10/10 - 8s - loss: 7.9642 - accuracy: 0.0844
Epoch 12/100
10/10 - 8s - loss: 8.0793 - accuracy: 0.0719
Epoch 13/100
10/10 - 8s - loss: 7.9559 - accuracy: 0.0969
Epoch 14/100
10/10 - 8s - loss: 7.7594 - accuracy: 0.0969
Epoch 15/100
10/10 - 8s - loss: 8.3687 - accuracy: 0.1094
Epoch 16/100
10/10 - 8s - loss: 7.0723 - accuracy: 0.1312
Epoch 17/100
10/10 - 8s - loss: 7.2359 - accuracy: 0.1156
Epoch 18/100
10/10 - 8s - loss: 6.9456 - accuracy: 0.1562
Epoch 19/100
10/10 - 8s - loss: 6.7341 - accuracy: 0.1781
Epoch 20/100
10/10 - 8s - loss: 6.0182 - accuracy: 0.2000
Epoch 21/100
10/10 - 8s - loss: 6.7215 - accuracy: 0.2000
Epoch 22/100
10/10 - 8s - loss: 6.5839 - accuracy: 0.2062
Epoch 23/100
10/10 - 8s - loss: 6.1088 - accuracy: 0.2500
Epoch 24/100
10/10 - 8s - loss: 5.9099 - accuracy: 0.2344
Epoch 25/100
10/10 - 8s - loss: 7.0490 - accuracy: 0.2031
Epoch 26/100
10

Epoch 49/100
10/10 - 8s - loss: 4.7394 - accuracy: 0.4344
Epoch 50/100
10/10 - 8s - loss: 3.9007 - accuracy: 0.5094
Epoch 51/100
10/10 - 8s - loss: 3.9125 - accuracy: 0.4906
Epoch 52/100
10/10 - 8s - loss: 4.3664 - accuracy: 0.4750
Epoch 53/100
10/10 - 8s - loss: 4.2187 - accuracy: 0.4563
Epoch 54/100
10/10 - 8s - loss: 3.8658 - accuracy: 0.4969
Epoch 55/100
10/10 - 8s - loss: 4.1464 - accuracy: 0.4750
Epoch 56/100
10/10 - 8s - loss: 4.2660 - accuracy: 0.5063
Epoch 57/100
10/10 - 8s - loss: 4.1816 - accuracy: 0.5094
Epoch 58/100
10/10 - 8s - loss: 3.8935 - accuracy: 0.5469
Epoch 59/100
10/10 - 8s - loss: 3.6310 - accuracy: 0.5781
Epoch 60/100
10/10 - 8s - loss: 3.7262 - accuracy: 0.5250
Epoch 61/100
10/10 - 8s - loss: 3.3173 - accuracy: 0.5688
Epoch 62/100
10/10 - 8s - loss: 4.1788 - accuracy: 0.4688
Epoch 63/100
10/10 - 8s - loss: 3.9080 - accuracy: 0.5281
Epoch 64/100
10/10 - 8s - loss: 5.1357 - accuracy: 0.4719
Epoch 65/100
10/10 - 8s - loss: 4.1551 - accuracy: 0.5031
Epoch 66/100
1

Epoch 89/100
10/10 - 8s - loss: 1.8041 - accuracy: 0.5250
Epoch 90/100
10/10 - 8s - loss: 1.5540 - accuracy: 0.5844
Epoch 91/100
10/10 - 8s - loss: 1.8151 - accuracy: 0.5531
Epoch 92/100
10/10 - 7s - loss: 1.8641 - accuracy: 0.5750
Epoch 93/100
10/10 - 8s - loss: 1.7351 - accuracy: 0.5656
Epoch 94/100
10/10 - 8s - loss: 1.7999 - accuracy: 0.5375
Epoch 95/100
10/10 - 8s - loss: 1.4807 - accuracy: 0.6156
Epoch 96/100
10/10 - 8s - loss: 1.6946 - accuracy: 0.5688
Epoch 97/100
10/10 - 8s - loss: 1.6175 - accuracy: 0.5844
Epoch 98/100
10/10 - 8s - loss: 1.6179 - accuracy: 0.5969
Epoch 99/100
10/10 - 8s - loss: 1.6432 - accuracy: 0.6187
Epoch 100/100
10/10 - 8s - loss: 1.4927 - accuracy: 0.6313
0.40796875227242707
2.6156194722652435
2.6156194722652435
0.40796875227242707
16 2 tanh 0.5 0.001
Epoch 1/100
10/10 - 9s - loss: 5.8456 - accuracy: 0.0031
Epoch 2/100
10/10 - 8s - loss: 5.9070 - accuracy: 0.0125
Epoch 3/100
10/10 - 8s - loss: 5.7563 - accuracy: 0.0094
Epoch 4/100
10/10 - 8s - loss: 5.5

10/10 - 8s - loss: 5.3446 - accuracy: 0.2937
Epoch 28/100
10/10 - 8s - loss: 5.5212 - accuracy: 0.3187
Epoch 29/100
10/10 - 8s - loss: 4.8366 - accuracy: 0.3438
Epoch 30/100
10/10 - 8s - loss: 5.4842 - accuracy: 0.3031
Epoch 31/100
10/10 - 8s - loss: 5.3190 - accuracy: 0.3187
Epoch 32/100
10/10 - 8s - loss: 5.5232 - accuracy: 0.3156
Epoch 33/100
10/10 - 8s - loss: 4.9990 - accuracy: 0.3281
Epoch 34/100
10/10 - 8s - loss: 5.4291 - accuracy: 0.3187
Epoch 35/100
10/10 - 8s - loss: 5.3055 - accuracy: 0.3063
Epoch 36/100
10/10 - 8s - loss: 4.6262 - accuracy: 0.3969
Epoch 37/100
10/10 - 8s - loss: 5.0766 - accuracy: 0.3906
Epoch 38/100
10/10 - 8s - loss: 5.3255 - accuracy: 0.4000
Epoch 39/100
10/10 - 8s - loss: 4.6753 - accuracy: 0.3844
Epoch 40/100
10/10 - 8s - loss: 4.8427 - accuracy: 0.3906
Epoch 41/100
10/10 - 8s - loss: 4.2835 - accuracy: 0.4781
Epoch 42/100
10/10 - 8s - loss: 4.5172 - accuracy: 0.4000
Epoch 43/100
10/10 - 8s - loss: 4.5412 - accuracy: 0.4125
Epoch 44/100
10/10 - 8s - l

Epoch 49/100
10/10 - 8s - loss: 2.1322 - accuracy: 0.4500
Epoch 50/100
10/10 - 8s - loss: 2.1089 - accuracy: 0.5125
Epoch 51/100
10/10 - 8s - loss: 1.8596 - accuracy: 0.5406
Epoch 52/100
10/10 - 8s - loss: 2.1308 - accuracy: 0.4812
Epoch 53/100
10/10 - 8s - loss: 2.0021 - accuracy: 0.5094
Epoch 54/100
10/10 - 8s - loss: 2.1186 - accuracy: 0.4719
Epoch 55/100
10/10 - 8s - loss: 2.0983 - accuracy: 0.4625
Epoch 56/100
10/10 - 8s - loss: 2.0732 - accuracy: 0.4812
Epoch 57/100
10/10 - 8s - loss: 1.8677 - accuracy: 0.5031
Epoch 58/100
10/10 - 8s - loss: 1.9947 - accuracy: 0.5094
Epoch 59/100
10/10 - 8s - loss: 1.9835 - accuracy: 0.5188
Epoch 60/100
10/10 - 8s - loss: 1.8796 - accuracy: 0.5469
Epoch 61/100
10/10 - 8s - loss: 2.0145 - accuracy: 0.4781
Epoch 62/100
10/10 - 8s - loss: 1.9204 - accuracy: 0.5125
Epoch 63/100
10/10 - 8s - loss: 1.8648 - accuracy: 0.5125
Epoch 64/100
10/10 - 8s - loss: 1.8391 - accuracy: 0.5156
Epoch 65/100
10/10 - 7s - loss: 2.1395 - accuracy: 0.4812
Epoch 66/100
1

Epoch 89/100
10/10 - 8s - loss: 1.6499 - accuracy: 0.5844
Epoch 90/100
10/10 - 8s - loss: 1.8230 - accuracy: 0.5781
Epoch 91/100
10/10 - 7s - loss: 1.7537 - accuracy: 0.5625
Epoch 92/100
10/10 - 8s - loss: 1.7924 - accuracy: 0.5625
Epoch 93/100
10/10 - 8s - loss: 1.7257 - accuracy: 0.5688
Epoch 94/100
10/10 - 8s - loss: 1.7456 - accuracy: 0.5406
Epoch 95/100
10/10 - 8s - loss: 1.6054 - accuracy: 0.6125
Epoch 96/100
10/10 - 8s - loss: 1.6506 - accuracy: 0.5969
Epoch 97/100
10/10 - 8s - loss: 1.7354 - accuracy: 0.5750
Epoch 98/100
10/10 - 7s - loss: 1.4386 - accuracy: 0.6281
Epoch 99/100
10/10 - 8s - loss: 1.7023 - accuracy: 0.5844
Epoch 100/100
10/10 - 8s - loss: 1.5443 - accuracy: 0.5844
0.39215625148965044
2.7071132230758668
2.7071132230758668
0.39215625148965044
256 2 tanh 0.5 0.001
Epoch 1/100
10/10 - 9s - loss: 9.2884 - accuracy: 0.0156
Epoch 2/100
10/10 - 8s - loss: 10.1729 - accuracy: 0.0125
Epoch 3/100
10/10 - 8s - loss: 9.3452 - accuracy: 0.0156
Epoch 4/100
10/10 - 8s - loss: 9

10/10 - 9s - loss: 4.8898 - accuracy: 0.0688
Epoch 28/100
10/10 - 9s - loss: 4.7409 - accuracy: 0.0469
Epoch 29/100
10/10 - 9s - loss: 4.8488 - accuracy: 0.0406
Epoch 30/100
10/10 - 9s - loss: 4.6471 - accuracy: 0.0812
Epoch 31/100
10/10 - 9s - loss: 4.5134 - accuracy: 0.0656
Epoch 32/100
10/10 - 9s - loss: 4.5337 - accuracy: 0.0938
Epoch 33/100
10/10 - 9s - loss: 4.4502 - accuracy: 0.0812
Epoch 34/100
10/10 - 9s - loss: 4.4306 - accuracy: 0.1094
Epoch 35/100
10/10 - 9s - loss: 4.2046 - accuracy: 0.0781
Epoch 36/100
10/10 - 9s - loss: 4.3325 - accuracy: 0.0969
Epoch 37/100
10/10 - 9s - loss: 4.1409 - accuracy: 0.1312
Epoch 38/100
10/10 - 9s - loss: 4.0626 - accuracy: 0.1063
Epoch 39/100
10/10 - 9s - loss: 4.2659 - accuracy: 0.1125
Epoch 40/100
10/10 - 9s - loss: 4.0754 - accuracy: 0.1094
Epoch 41/100
10/10 - 9s - loss: 4.0073 - accuracy: 0.1375
Epoch 42/100
10/10 - 9s - loss: 4.1681 - accuracy: 0.1031
Epoch 43/100
10/10 - 9s - loss: 4.0388 - accuracy: 0.1312
Epoch 44/100
10/10 - 9s - l

Epoch 67/100
10/10 - 8s - loss: 2.6640 - accuracy: 0.3688
Epoch 68/100
10/10 - 8s - loss: 2.6311 - accuracy: 0.3906
Epoch 69/100
10/10 - 8s - loss: 2.8455 - accuracy: 0.3063
Epoch 70/100
10/10 - 8s - loss: 2.8082 - accuracy: 0.3313
Epoch 71/100
10/10 - 8s - loss: 2.8170 - accuracy: 0.3094
Epoch 72/100
10/10 - 8s - loss: 2.5342 - accuracy: 0.3750
Epoch 73/100
10/10 - 8s - loss: 2.7204 - accuracy: 0.3281
Epoch 74/100
10/10 - 8s - loss: 2.7478 - accuracy: 0.3344
Epoch 75/100
10/10 - 8s - loss: 2.7259 - accuracy: 0.3344
Epoch 76/100
10/10 - 8s - loss: 2.4296 - accuracy: 0.3906
Epoch 77/100
10/10 - 8s - loss: 2.8030 - accuracy: 0.3438
Epoch 78/100
10/10 - 8s - loss: 2.4611 - accuracy: 0.4000
Epoch 79/100
10/10 - 8s - loss: 2.6813 - accuracy: 0.3656
Epoch 80/100
10/10 - 8s - loss: 2.4057 - accuracy: 0.3594
Epoch 81/100
10/10 - 8s - loss: 2.7243 - accuracy: 0.3500
Epoch 82/100
10/10 - 8s - loss: 2.5194 - accuracy: 0.4125
Epoch 83/100
10/10 - 8s - loss: 2.5475 - accuracy: 0.3625
Epoch 84/100
1

10/10 - 8s - loss: 4.2148 - accuracy: 0.5719
Epoch 89/100
10/10 - 8s - loss: 4.0053 - accuracy: 0.5750
Epoch 90/100
10/10 - 8s - loss: 4.0768 - accuracy: 0.5656
Epoch 91/100
10/10 - 9s - loss: 4.1587 - accuracy: 0.5344
Epoch 92/100
10/10 - 8s - loss: 3.9892 - accuracy: 0.5875
Epoch 93/100
10/10 - 8s - loss: 4.7197 - accuracy: 0.5625
Epoch 94/100
10/10 - 8s - loss: 4.6269 - accuracy: 0.5188
Epoch 95/100
10/10 - 8s - loss: 4.4725 - accuracy: 0.5469
Epoch 96/100
10/10 - 8s - loss: 4.1494 - accuracy: 0.5938
Epoch 97/100
10/10 - 8s - loss: 3.9895 - accuracy: 0.6031
Epoch 98/100
10/10 - 8s - loss: 3.5783 - accuracy: 0.6375
Epoch 99/100
10/10 - 8s - loss: 3.7660 - accuracy: 0.6406
Epoch 100/100
10/10 - 8s - loss: 4.2581 - accuracy: 0.5531
0.39914727716939524
5.205491037368774
5.205491037368774
0.39914727716939524
16 2 tanh 0.5 0.001
Epoch 1/100
10/10 - 9s - loss: 5.8709 - accuracy: 0.0094
Epoch 2/100
10/10 - 8s - loss: 5.9307 - accuracy: 0.0031
Epoch 3/100
10/10 - 8s - loss: 5.8262 - accuracy

Epoch 27/100
10/10 - 8s - loss: 2.9598 - accuracy: 0.3094
Epoch 28/100
10/10 - 8s - loss: 2.7985 - accuracy: 0.3094
Epoch 29/100
10/10 - 8s - loss: 2.8536 - accuracy: 0.2937
Epoch 30/100
10/10 - 8s - loss: 2.9939 - accuracy: 0.3313
Epoch 31/100
10/10 - 8s - loss: 2.5889 - accuracy: 0.3875
Epoch 32/100
10/10 - 8s - loss: 2.7038 - accuracy: 0.3344
Epoch 33/100
10/10 - 8s - loss: 2.6864 - accuracy: 0.3656
Epoch 34/100
10/10 - 8s - loss: 2.5287 - accuracy: 0.4250
Epoch 35/100
10/10 - 8s - loss: 2.5498 - accuracy: 0.3906
Epoch 36/100
10/10 - 8s - loss: 2.4378 - accuracy: 0.4031
Epoch 37/100
10/10 - 8s - loss: 2.5970 - accuracy: 0.4125
Epoch 38/100
10/10 - 8s - loss: 2.6879 - accuracy: 0.3406
Epoch 39/100
10/10 - 8s - loss: 2.3862 - accuracy: 0.4281
Epoch 40/100
10/10 - 8s - loss: 2.5504 - accuracy: 0.3906
Epoch 41/100
10/10 - 8s - loss: 2.4128 - accuracy: 0.3719
Epoch 42/100
10/10 - 8s - loss: 2.3963 - accuracy: 0.3750
Epoch 43/100
10/10 - 8s - loss: 2.5306 - accuracy: 0.4219
Epoch 44/100
1

Epoch 67/100
10/10 - 8s - loss: 1.9264 - accuracy: 0.4969
Epoch 68/100
10/10 - 8s - loss: 1.9152 - accuracy: 0.5188
Epoch 69/100
10/10 - 8s - loss: 1.7778 - accuracy: 0.5188
Epoch 70/100
10/10 - 8s - loss: 1.7182 - accuracy: 0.5312
Epoch 71/100
10/10 - 8s - loss: 1.9139 - accuracy: 0.5125
Epoch 72/100
10/10 - 8s - loss: 1.9285 - accuracy: 0.5281
Epoch 73/100
10/10 - 8s - loss: 1.9069 - accuracy: 0.5094
Epoch 74/100
10/10 - 8s - loss: 1.5310 - accuracy: 0.6313
Epoch 75/100
10/10 - 8s - loss: 1.7682 - accuracy: 0.5688
Epoch 76/100
10/10 - 8s - loss: 1.8261 - accuracy: 0.5375
Epoch 77/100
10/10 - 8s - loss: 1.7971 - accuracy: 0.5500
Epoch 78/100
10/10 - 8s - loss: 1.7715 - accuracy: 0.5594
Epoch 79/100
10/10 - 8s - loss: 1.7355 - accuracy: 0.5188
Epoch 80/100
10/10 - 8s - loss: 1.6387 - accuracy: 0.5656
Epoch 81/100
10/10 - 8s - loss: 1.9792 - accuracy: 0.5094
Epoch 82/100
10/10 - 8s - loss: 1.6781 - accuracy: 0.5813
Epoch 83/100
10/10 - 8s - loss: 1.8946 - accuracy: 0.5594
Epoch 84/100
1

Epoch 1/100
10/10 - 9s - loss: 5.9803 - accuracy: 0.0031
Epoch 2/100
10/10 - 8s - loss: 5.8900 - accuracy: 0.0063
Epoch 3/100
10/10 - 8s - loss: 5.8091 - accuracy: 0.0125
Epoch 4/100
10/10 - 8s - loss: 5.6898 - accuracy: 0.0156
Epoch 5/100
10/10 - 8s - loss: 5.5077 - accuracy: 0.0312
Epoch 6/100
10/10 - 8s - loss: 5.3593 - accuracy: 0.0219
Epoch 7/100
10/10 - 8s - loss: 5.1359 - accuracy: 0.0750
Epoch 8/100
10/10 - 8s - loss: 5.0004 - accuracy: 0.0750
Epoch 9/100
10/10 - 8s - loss: 4.8798 - accuracy: 0.0812
Epoch 10/100
10/10 - 8s - loss: 4.7079 - accuracy: 0.0906
Epoch 11/100
10/10 - 8s - loss: 4.5784 - accuracy: 0.1187
Epoch 12/100
10/10 - 8s - loss: 4.4827 - accuracy: 0.1031
Epoch 13/100
10/10 - 8s - loss: 4.1243 - accuracy: 0.1688
Epoch 14/100
10/10 - 8s - loss: 4.1355 - accuracy: 0.1344
Epoch 15/100
10/10 - 8s - loss: 3.9696 - accuracy: 0.1625
Epoch 16/100
10/10 - 8s - loss: 3.8212 - accuracy: 0.1688
Epoch 17/100
10/10 - 8s - loss: 3.7773 - accuracy: 0.1875
Epoch 18/100
10/10 - 8s

Epoch 41/100
10/10 - 7s - loss: 4.6315 - accuracy: 0.4187
Epoch 42/100
10/10 - 7s - loss: 4.8000 - accuracy: 0.4406
Epoch 43/100
10/10 - 7s - loss: 4.6104 - accuracy: 0.4500
Epoch 44/100
10/10 - 7s - loss: 4.4034 - accuracy: 0.4406
Epoch 45/100
10/10 - 8s - loss: 4.3551 - accuracy: 0.4625
Epoch 46/100
10/10 - 7s - loss: 4.7718 - accuracy: 0.4313
Epoch 47/100
10/10 - 10s - loss: 4.3426 - accuracy: 0.4375
Epoch 48/100
10/10 - 8s - loss: 4.8993 - accuracy: 0.4031
Epoch 49/100
10/10 - 7s - loss: 4.7300 - accuracy: 0.4875
Epoch 50/100
10/10 - 7s - loss: 4.0325 - accuracy: 0.5094
Epoch 51/100
10/10 - 7s - loss: 4.6373 - accuracy: 0.4563
Epoch 52/100
10/10 - 7s - loss: 4.2745 - accuracy: 0.4563
Epoch 53/100
10/10 - 7s - loss: 4.5924 - accuracy: 0.4656
Epoch 54/100
10/10 - 7s - loss: 4.3151 - accuracy: 0.4563
Epoch 55/100
10/10 - 7s - loss: 4.0435 - accuracy: 0.4719
Epoch 56/100
10/10 - 7s - loss: 4.2004 - accuracy: 0.4906
Epoch 57/100
10/10 - 7s - loss: 4.0361 - accuracy: 0.4906
Epoch 58/100


Epoch 81/100
10/10 - 7s - loss: 1.5483 - accuracy: 0.5906
Epoch 82/100
10/10 - 6s - loss: 1.7840 - accuracy: 0.5375
Epoch 83/100
10/10 - 6s - loss: 1.6649 - accuracy: 0.5813
Epoch 84/100
10/10 - 6s - loss: 1.6217 - accuracy: 0.5813
Epoch 85/100
10/10 - 6s - loss: 1.7935 - accuracy: 0.5469
Epoch 86/100
10/10 - 6s - loss: 1.6218 - accuracy: 0.6187
Epoch 87/100
10/10 - 6s - loss: 1.6631 - accuracy: 0.5750
Epoch 88/100
10/10 - 6s - loss: 1.6577 - accuracy: 0.5781
Epoch 89/100
10/10 - 6s - loss: 1.5855 - accuracy: 0.6156
Epoch 90/100
10/10 - 6s - loss: 1.7463 - accuracy: 0.5562
Epoch 91/100
10/10 - 6s - loss: 1.6831 - accuracy: 0.5688
Epoch 92/100
10/10 - 7s - loss: 1.7760 - accuracy: 0.5688
Epoch 93/100
10/10 - 6s - loss: 1.5718 - accuracy: 0.5969
Epoch 94/100
10/10 - 6s - loss: 1.5625 - accuracy: 0.5969
Epoch 95/100
10/10 - 7s - loss: 1.4270 - accuracy: 0.6281
Epoch 96/100
10/10 - 6s - loss: 1.5299 - accuracy: 0.6219
Epoch 97/100
10/10 - 6s - loss: 1.7266 - accuracy: 0.5531
Epoch 98/100
1

10/10 - 7s - loss: 6.3419 - accuracy: 0.2156
Epoch 20/100
10/10 - 7s - loss: 6.9560 - accuracy: 0.1562
Epoch 21/100
10/10 - 7s - loss: 6.3394 - accuracy: 0.2281
Epoch 22/100
10/10 - 7s - loss: 6.6069 - accuracy: 0.2406
Epoch 23/100
10/10 - 7s - loss: 6.9550 - accuracy: 0.1844
Epoch 24/100
10/10 - 7s - loss: 5.9898 - accuracy: 0.2313
Epoch 25/100
10/10 - 7s - loss: 6.2171 - accuracy: 0.2406
Epoch 26/100
10/10 - 7s - loss: 6.4089 - accuracy: 0.2406
Epoch 27/100
10/10 - 7s - loss: 6.1195 - accuracy: 0.2688
Epoch 28/100
10/10 - 7s - loss: 5.7646 - accuracy: 0.2781
Epoch 29/100
10/10 - 7s - loss: 5.8573 - accuracy: 0.2906
Epoch 30/100
10/10 - 7s - loss: 5.6613 - accuracy: 0.2875
Epoch 31/100
10/10 - 7s - loss: 5.3367 - accuracy: 0.3469
Epoch 32/100
10/10 - 7s - loss: 6.0269 - accuracy: 0.3406
Epoch 33/100
10/10 - 7s - loss: 4.9784 - accuracy: 0.3438
Epoch 34/100
10/10 - 7s - loss: 5.3799 - accuracy: 0.3063
Epoch 35/100
10/10 - 7s - loss: 5.1696 - accuracy: 0.3500
Epoch 36/100
10/10 - 7s - l

Epoch 41/100
10/10 - 6s - loss: 2.2653 - accuracy: 0.4344
Epoch 42/100
10/10 - 6s - loss: 2.3925 - accuracy: 0.3938
Epoch 43/100
10/10 - 7s - loss: 2.2888 - accuracy: 0.4344
Epoch 44/100
10/10 - 6s - loss: 2.3634 - accuracy: 0.3875
Epoch 45/100
10/10 - 6s - loss: 2.0492 - accuracy: 0.4969
Epoch 46/100
10/10 - 6s - loss: 2.2873 - accuracy: 0.4594
Epoch 47/100
10/10 - 6s - loss: 2.1531 - accuracy: 0.4875
Epoch 48/100
10/10 - 6s - loss: 2.0934 - accuracy: 0.4781
Epoch 49/100
10/10 - 6s - loss: 2.0972 - accuracy: 0.4656
Epoch 50/100
10/10 - 6s - loss: 2.2440 - accuracy: 0.4437
Epoch 51/100
10/10 - 6s - loss: 2.1240 - accuracy: 0.4781
Epoch 52/100
10/10 - 6s - loss: 2.1878 - accuracy: 0.4563
Epoch 53/100
10/10 - 6s - loss: 2.1011 - accuracy: 0.4812
Epoch 54/100
10/10 - 6s - loss: 2.1259 - accuracy: 0.4719
Epoch 55/100
10/10 - 6s - loss: 2.0378 - accuracy: 0.4688
Epoch 56/100
10/10 - 6s - loss: 2.0607 - accuracy: 0.4688
Epoch 57/100
10/10 - 7s - loss: 1.8666 - accuracy: 0.5437
Epoch 58/100
1

Epoch 81/100
10/10 - 7s - loss: 1.9769 - accuracy: 0.5063
Epoch 82/100
10/10 - 6s - loss: 1.6424 - accuracy: 0.5938
Epoch 83/100
10/10 - 6s - loss: 1.6847 - accuracy: 0.5625
Epoch 84/100
10/10 - 6s - loss: 1.7165 - accuracy: 0.5531
Epoch 85/100
10/10 - 6s - loss: 1.7890 - accuracy: 0.5500
Epoch 86/100
10/10 - 6s - loss: 1.5712 - accuracy: 0.6187
Epoch 87/100
10/10 - 6s - loss: 1.5532 - accuracy: 0.5656
Epoch 88/100
10/10 - 6s - loss: 1.7473 - accuracy: 0.5594
Epoch 89/100
10/10 - 6s - loss: 1.7723 - accuracy: 0.5688
Epoch 90/100
10/10 - 6s - loss: 1.7031 - accuracy: 0.5844
Epoch 91/100
10/10 - 6s - loss: 1.3189 - accuracy: 0.6531
Epoch 92/100
10/10 - 7s - loss: 1.5098 - accuracy: 0.5969
Epoch 93/100
10/10 - 6s - loss: 1.4643 - accuracy: 0.6156
Epoch 94/100
10/10 - 6s - loss: 1.4636 - accuracy: 0.6156
Epoch 95/100
10/10 - 7s - loss: 1.7916 - accuracy: 0.5437
Epoch 96/100
10/10 - 6s - loss: 1.5382 - accuracy: 0.5844
Epoch 97/100
10/10 - 6s - loss: 1.6740 - accuracy: 0.5531
Epoch 98/100
1

10/10 - 7s - loss: 6.3922 - accuracy: 0.2000
Epoch 20/100
10/10 - 7s - loss: 6.3951 - accuracy: 0.2031
Epoch 21/100
10/10 - 7s - loss: 6.2994 - accuracy: 0.2250
Epoch 22/100
10/10 - 7s - loss: 6.5265 - accuracy: 0.2406
Epoch 23/100
10/10 - 7s - loss: 6.3589 - accuracy: 0.2344
Epoch 24/100
10/10 - 7s - loss: 5.7596 - accuracy: 0.2750
Epoch 25/100
10/10 - 7s - loss: 6.1063 - accuracy: 0.2406
Epoch 26/100
10/10 - 7s - loss: 5.5581 - accuracy: 0.2844
Epoch 27/100
10/10 - 7s - loss: 6.2602 - accuracy: 0.2844
Epoch 28/100
10/10 - 7s - loss: 6.2975 - accuracy: 0.3031
Epoch 29/100
10/10 - 7s - loss: 6.1099 - accuracy: 0.3219
Epoch 30/100
10/10 - 7s - loss: 6.0200 - accuracy: 0.2344
Epoch 31/100
10/10 - 7s - loss: 5.3396 - accuracy: 0.3187
Epoch 32/100
10/10 - 7s - loss: 5.2659 - accuracy: 0.3250
Epoch 33/100
10/10 - 7s - loss: 5.1876 - accuracy: 0.3281
Epoch 34/100
10/10 - 7s - loss: 5.7661 - accuracy: 0.3187
Epoch 35/100
10/10 - 7s - loss: 5.0100 - accuracy: 0.3625
Epoch 36/100
10/10 - 7s - l

Epoch 59/100
10/10 - 7s - loss: 1.9509 - accuracy: 0.5063
Epoch 60/100
10/10 - 7s - loss: 1.9204 - accuracy: 0.5250
Epoch 61/100
10/10 - 7s - loss: 1.7600 - accuracy: 0.5531
Epoch 62/100
10/10 - 6s - loss: 1.9445 - accuracy: 0.5063
Epoch 63/100
10/10 - 6s - loss: 1.9465 - accuracy: 0.5250
Epoch 64/100
10/10 - 7s - loss: 2.0366 - accuracy: 0.4812
Epoch 65/100
10/10 - 6s - loss: 1.7650 - accuracy: 0.5719
Epoch 66/100
10/10 - 7s - loss: 1.8107 - accuracy: 0.5031
Epoch 67/100
10/10 - 7s - loss: 1.5717 - accuracy: 0.6156
Epoch 68/100
10/10 - 6s - loss: 1.7349 - accuracy: 0.5625
Epoch 69/100
10/10 - 6s - loss: 1.7696 - accuracy: 0.5469
Epoch 70/100
10/10 - 7s - loss: 1.8292 - accuracy: 0.5750
Epoch 71/100
10/10 - 6s - loss: 1.5587 - accuracy: 0.5719
Epoch 72/100
10/10 - 7s - loss: 2.0983 - accuracy: 0.5156
Epoch 73/100
10/10 - 7s - loss: 1.7764 - accuracy: 0.5813
Epoch 74/100
10/10 - 6s - loss: 1.7923 - accuracy: 0.5562
Epoch 75/100
10/10 - 7s - loss: 1.9951 - accuracy: 0.4750
Epoch 76/100
1

Epoch 81/100
10/10 - 6s - loss: 1.7731 - accuracy: 0.5375
Epoch 82/100
10/10 - 6s - loss: 1.7831 - accuracy: 0.5469
Epoch 83/100
10/10 - 6s - loss: 1.9093 - accuracy: 0.5094
Epoch 84/100
10/10 - 6s - loss: 1.7445 - accuracy: 0.5594
Epoch 85/100
10/10 - 6s - loss: 1.8066 - accuracy: 0.5437
Epoch 86/100
10/10 - 6s - loss: 1.8419 - accuracy: 0.5656
Epoch 87/100
10/10 - 6s - loss: 1.6577 - accuracy: 0.5844
Epoch 88/100
10/10 - 7s - loss: 1.7500 - accuracy: 0.5875
Epoch 89/100
10/10 - 6s - loss: 1.6174 - accuracy: 0.5719
Epoch 90/100
10/10 - 6s - loss: 1.5037 - accuracy: 0.6062
Epoch 91/100
10/10 - 7s - loss: 1.6376 - accuracy: 0.5906
Epoch 92/100
10/10 - 6s - loss: 1.6602 - accuracy: 0.5656
Epoch 93/100
10/10 - 6s - loss: 1.6689 - accuracy: 0.5938
Epoch 94/100
10/10 - 6s - loss: 1.4207 - accuracy: 0.6375
Epoch 95/100
10/10 - 6s - loss: 1.5571 - accuracy: 0.5781
Epoch 96/100
10/10 - 6s - loss: 1.6249 - accuracy: 0.5562
Epoch 97/100
10/10 - 6s - loss: 1.4507 - accuracy: 0.6250
Epoch 98/100
1

10/10 - 7s - loss: 3.7641 - accuracy: 0.1969
Epoch 20/100
10/10 - 6s - loss: 3.7601 - accuracy: 0.1906
Epoch 21/100
10/10 - 6s - loss: 3.5793 - accuracy: 0.2250
Epoch 22/100
10/10 - 7s - loss: 3.3507 - accuracy: 0.2469
Epoch 23/100
10/10 - 6s - loss: 3.1215 - accuracy: 0.2969
Epoch 24/100
10/10 - 6s - loss: 3.1192 - accuracy: 0.2812
Epoch 25/100
10/10 - 7s - loss: 3.1627 - accuracy: 0.2375
Epoch 26/100
10/10 - 6s - loss: 2.9047 - accuracy: 0.3094
Epoch 27/100
10/10 - 6s - loss: 3.1239 - accuracy: 0.3125
Epoch 28/100
10/10 - 6s - loss: 2.8273 - accuracy: 0.3250
Epoch 29/100
10/10 - 6s - loss: 2.8341 - accuracy: 0.3313
Epoch 30/100
10/10 - 7s - loss: 2.7919 - accuracy: 0.3187
Epoch 31/100
10/10 - 6s - loss: 2.7999 - accuracy: 0.3031
Epoch 32/100
10/10 - 6s - loss: 2.8085 - accuracy: 0.3125
Epoch 33/100
10/10 - 6s - loss: 2.6903 - accuracy: 0.3594
Epoch 34/100
10/10 - 6s - loss: 2.4782 - accuracy: 0.3938
Epoch 35/100
10/10 - 6s - loss: 2.5145 - accuracy: 0.3750
Epoch 36/100
10/10 - 7s - l

Epoch 59/100
10/10 - 6s - loss: 2.1855 - accuracy: 0.4437
Epoch 60/100
10/10 - 6s - loss: 2.0033 - accuracy: 0.5063
Epoch 61/100
10/10 - 6s - loss: 2.0663 - accuracy: 0.4906
Epoch 62/100
10/10 - 6s - loss: 1.6775 - accuracy: 0.5875
Epoch 63/100
10/10 - 6s - loss: 2.1058 - accuracy: 0.4875
Epoch 64/100
10/10 - 6s - loss: 1.8207 - accuracy: 0.5500
Epoch 65/100
10/10 - 6s - loss: 1.9067 - accuracy: 0.5156
Epoch 66/100
10/10 - 6s - loss: 1.8433 - accuracy: 0.5344
Epoch 67/100
10/10 - 7s - loss: 1.7524 - accuracy: 0.5562
Epoch 68/100
10/10 - 6s - loss: 1.8399 - accuracy: 0.5594
Epoch 69/100
10/10 - 6s - loss: 1.6527 - accuracy: 0.5562
Epoch 70/100
10/10 - 7s - loss: 1.7779 - accuracy: 0.5437
Epoch 71/100
10/10 - 6s - loss: 1.9661 - accuracy: 0.5156
Epoch 72/100
10/10 - 6s - loss: 1.6162 - accuracy: 0.5781
Epoch 73/100
10/10 - 6s - loss: 1.7495 - accuracy: 0.5625
Epoch 74/100
10/10 - 6s - loss: 1.7813 - accuracy: 0.5469
Epoch 75/100
10/10 - 6s - loss: 1.8822 - accuracy: 0.5281
Epoch 76/100
1

Epoch 99/100
10/10 - 6s - loss: 1.3382 - accuracy: 0.6438
Epoch 100/100
10/10 - 7s - loss: 1.3859 - accuracy: 0.6406
0.409781250001397
2.6004033851623536
2.6004033851623536
0.409781250001397
Selected Ranked Chromosomes:
[16, 2, 'tanh', 0.5, 0.001, 2.6004033851623536, 0.409781250001397]
[16, 2, 'tanh', 0.5, 0.001, 2.613929785490036, 0.40974999997066336]
[16, 2, 'tanh', 0.5, 0.001, 2.614882788658142, 0.4069167682295665]
Child 1: [16, 2, 'tanh', 0.5, 0.001, 2.6004033851623536, 0.409781250001397] child 2: [16, 2, 'tanh', 0.5, 0.001, 2.613929785490036, 0.40974999997066336]
[16, 2, 'tanh', 0.5, 0.001, 2.6004033851623536, 0.409781250001397]
[16, 2, 'tanh', 0.5, 0.001, 2.613929785490036, 0.40974999997066336]
rate: 0.02105898445900245
length: 5
index: 4
New Child9
[16, 2, 'tanh', 0.5, 0.1, 2.6004033851623536, 0.409781250001397]
[16, 2, 'tanh', 0.5, 0.01, 2.613929785490036, 0.40974999997066336]
New Generation 10
[16, 2, 'tanh', 0.5, 0.001, 2.6004033851623536, 0.409781250001397]
[16, 2, 'tanh', 0

In [6]:
print(population)

[[16, 2, 'tanh', 0.5, 0.001, 2.6004033851623536, 0.409781250001397], [16, 2, 'tanh', 0.5, 0.001, 2.613929785490036, 0.40974999997066336], [16, 2, 'tanh', 0.5, 0.001, 2.614882788658142, 0.4069167682295665], [16, 2, 'tanh', 0.5, 0.1, 2.6004033851623536, 0.409781250001397], [16, 2, 'tanh', 0.5, 0.01, 2.613929785490036, 0.40974999997066336]]


# Tanto fazia mas o cromossoma que escolhi foi:

[16, 2, 'tanh', 0.5, 0.01, 2.6004033851623536, 0.409781250001397]

Será corrido, no ficheiro "CNN-with-GA.ipynb", o model.fit com estes parâmetros. 
Pequenas notas:
- O que difere entre os cromossomas resultantes foi apenas o LR, sendo assim escolhi o que tinha maior accuracy (que é quase insignificante)
- No entanto, no modelo CNN do ficheiro mencionado acima, coloquei uma callback chamada ReduceLROnPlateau na função de fitness que vai reduzindo o LR quando a accuracy para de melhorar. O LR passou de 0.01 -> 0.0005. É uma grande diferença mas podia ter usado um dos outras cromossomas resultantes (que têm todos praticamente os mesmos valores de loss e accuracy) com LR de 0.001 que a diferença já seria menor.
- Também consigo concluir que, apesar de ter colocado valores como 0.0005 nas possibilidades para as combinações dos parâmetros, este não foi escolhido pelo algoritmo inicialmente, o que leva a pensar que, dada outra combinação de valores iniciais, o meu resultado poderia ter sido melhor, não só para o LR como para todos os outros.